exonum-java-proofs/src/test/java/com/exonum/binding/storage/proofs/map/flat/UncheckedFlatMapProofTest.java

```diff
@@ -0,0 +1,254 @@
+            emptyList(),
+            singletonList(mapEntry),
+            emptyList());
+    CheckedMapProof checkedMapProof = uncheckedFlatMapProof.check();
+
+    assertThat(checkedMapProof.getRootHash(), equalTo(expectedRootHash));
+    assertTrue(checkedMapProof.compareWithRootHash(expectedRootHash));
+
+    assertThat(checkedMapProof.getEntries(), equalTo(singletonList(mapEntry)));
+    assertTrue(checkedMapProof.containsKey(key));
+    assertThat(checkedMapProof.get(key), equalTo(value));
+  }
+
+  @Test
+  public void mapProofWithEqualEntriesOrderShouldBeInvalid() {
+    DbKey firstDbKey = DbKeyTestUtils.branchKeyFromPrefix("101100");
+    DbKey secondKey = DbKeyTestUtils.branchKeyFromPrefix("101100");
+    List<MapProofEntry> entries = Arrays.asList(
+        createMapProofEntry(firstDbKey),
+        createMapProofEntry(secondKey)
+    );
+    UncheckedMapProof uncheckedFlatMapProof =
+        new UncheckedFlatMapProof(entries, emptyList(), emptyList());
+
+    CheckedMapProof checkedMapProof = uncheckedFlatMapProof.check();
+    assertThat(checkedMapProof.getStatus(), equalTo(ProofStatus.DUPLICATE_PATH));
+  }
+
+  @Test
+  public void mapProofWithoutEntriesShouldBeValid() {
+    UncheckedMapProof uncheckedFlatMapProof =
+        new UncheckedFlatMapProof(
+            emptyList(), emptyList(), emptyList());
+
+    CheckedMapProof checkedMapProof = uncheckedFlatMapProof.check();
+    assertThat(checkedMapProof.getStatus(), equalTo(ProofStatus.CORRECT));
+  }
+
+  @Test
+  public void mapProofWithAbsentKeyShouldBeCorrect() {
+    DbKey firstDbKey = DbKeyTestUtils.branchKeyFromPrefix("101100");
+    byte[] valueKey = DbKeyTestUtils.keyFromString("101110");
+    byte[] absentKey = DbKeyTestUtils.keyFromString("101111");
+
+    UncheckedMapProof uncheckedFlatMapProof =
+        new UncheckedFlatMapProof(
+            singletonList(createMapProofEntry(firstDbKey)),
+            singletonList(createMapEntry(valueKey, FIRST_VALUE)),
+            singletonList(absentKey));
+
+    CheckedMapProof checkedMapProof = uncheckedFlatMapProof.check();
+    assertThat(checkedMapProof.getStatus(), equalTo(ProofStatus.CORRECT));
+  }
+
+  @Test
+  public void mapProofWithInvalidOrderShouldBeIncorrect() {
+    DbKey firstDbKey = DbKeyTestUtils.branchKeyFromPrefix("10");
+    DbKey secondDbKey = DbKeyTestUtils.branchKeyFromPrefix("01");
+
+    UncheckedMapProof uncheckedFlatMapProof =
+        new UncheckedFlatMapProof(
+            Arrays.asList(createMapProofEntry(firstDbKey), createMapProofEntry(secondDbKey)),
+            emptyList(),
+            emptyList());
+
+    CheckedMapProof checkedMapProof = uncheckedFlatMapProof.check();
+    assertThat(checkedMapProof.getStatus(), equalTo(ProofStatus.INVALID_ORDER));
+  }
+
+  @Test
+  public void mapProofWithSingleBranchProofEntryShouldBeInvalid() {
+    DbKey firstDbKey = DbKeyTestUtils.branchKeyFromPrefix("1011111");
+    byte[] absentKey = DbKeyTestUtils.keyFromString("101111");
```

Comment:
I'm not completely sure, but allowing to specify missing keys as arbitrary byte sequences (instead of only "full" keys corresponding to leaves) looks weird. OTOH, I suppose there are no reasons for it to not work, provided that the length of the key is checked to be less or equal to the length of a leaf key.

Reply:
As I remember, `keyFromString` must create a proper 32-byte key, that has the given _bit_ prefix (and an all-zero tail). It used to be a private method which name made perfect sense in that context — but in a wider context it might be confusing.

I am not sure though that we check the size of keys explicitly ­— but we certainly do _implicitly_ here (an exception will occur if it is not 32-byte long):
https://github.com/exonum/exonum-java-binding/pull/250/files/02985fb9adabfd387830e8e4dbdcd78e365dc26f#diff-2bbe03c0a0570e463c254c07245b4bd3R123